File: code/126_127.py
#word ladder I
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordset, wordgraph, queue = set(wordList), defaultdict(set), deque([(beginWord, 1)])
        if endWord not in wordset:
            return 0
        while queue:
            word, step = queue.popleft()
            if word not in wordgraph:
                for i in range(len(word)):
                    for ch in string.ascii_lowercase:
                        newword = word[:i] + ch + word[i + 1:]
                        if ch == word[i]:
                            continue
                        if newword in wordset:
                            wordgraph[word].add(newword)
            if word == endWord:
                return step
            else:
                for newword in wordgraph[word]:
                    queue.append((newword, step + 1))
                wordgraph[word] = set()
        return 0
#word ladder II
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        wordset, queue, worddis, wordgraph = set(wordList), deque([(beginWord, 1)]), {beginWord : 1}, defaultdict(set)
        res, length = [], inf
        if endWord not in wordset:
            return []
        while queue and queue[0][1] < length:
            word, step = queue.popleft()
            if word not in wordgraph:
                for i in range(len(word)):
                    for ch in string.ascii_lowercase:
                        if ch == word[i]:
                            continue
                        newword = word[:i] + ch + word[i + 1:]
                        if newword in wordset:
                            wordgraph[word].add(newword)
            for nextword in wordgraph[word]:
                if nextword not in worddis:
                    worddis[nextword] = step + 1
                elif nextword in worddis and worddis[nextword] < step + 1:
                    continue
                if nextword == endWord:
                    length = step + 1
                else:
                    queue.append((nextword, step + 1))
        if length == inf:
            return []

        def backtracking(path, curword):
            if worddis[curword] < len(path):
                return
            if curword == endWord:
                res.append(path[:])
                return
            for nextword in wordgraph[curword]:
                path.append(nextword)
                backtracking(path, nextword)
                path.pop()

        backtracking([beginWord], beginWord)

        return res

# the method 2 below is much lower than the method 1;
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        wordList = set(wordList)
        if endWord not in wordList:
            return []

        path_q = deque([[beginWord]])
        visited = set([beginWord])
        visit_pending = set()
        res = []
        level = 1
        length = len(beginWord)
        while path_q:
            curr_path = path_q.popleft()
            if len(curr_path) > level:
                level += 1
                visited.update(visit_pending)
                visit_pending = set()
            word = curr_path[-1]
            for i in range(length):
                for letter in 'abcdefghijklmnopqrstuvwxyz':
                    new_word = word[:i] + letter + word[i+1:]
                    if new_word in wordList and new_word not in visited:
                        new_path = curr_path + [new_word]
                        visit_pending.add(new_word)
                        if new_word == endWord:
                            res.append(new_path)
                        else:
                            path_q.append(new_path)

        return res

File: code/126_127.java
// 127
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord) || beginWord.equals(endWord)) return 0;
        wordSet.remove(beginWord);
        Queue<String> queue = new LinkedList<>();
        int wordSize = beginWord.length();
        queue.add(beginWord);
        int steps = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int k = 0; k < size; k++) {
                String word = queue.poll();
                if (word.equals(endWord)) return steps+1;
                char[] letters = word.toCharArray(); // convert string to array
                // It is faster to generate a new word by converting to array and converting back, rather than generating a new string by string slice.
                for (int i = 0; i < wordSize; i++) {
                    char original = letters[i];
                    for (char j = 'a'; j <= 'z'; j++) {
                        letters[i] = j;
                        String newWord = new String(letters); // convert array to string
                        if (!wordSet.contains(newWord)) continue;
                        wordSet.remove(newWord);
                        queue.add(newWord);
                    }
                    letters[i] = original;
                }
            }
            steps++;
        }
        return 0;
    }
}

// 127 method 2: Two-end BFS
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> words = new HashSet<>(wordList);
        if(!words.contains(endWord)) return 0;
        Set<String> beginSet = new HashSet<>();
        Set<String> endSet = new HashSet<>();
        beginSet.add(beginWord);
        endSet.add(endWord);
        int length = 1;
        while(!beginSet.isEmpty() && !endSet.isEmpty()){
            if(beginSet.size()>endSet.size()){
                Set<String> temp = beginSet;
                beginSet = endSet;
                endSet = temp;
            }
            Set<String> newBeginSet = new HashSet<>();
            for(String word: beginSet){
                List<String> neighbors = neighbors(word);
                for(String neigh:neighbors){
                    if(endSet.contains(neigh)) return length+1;
                    if(words.contains(neigh)){
                        newBeginSet.add(neigh);
                        words.remove(neigh);
                    }
                }
            }
            beginSet = newBeginSet;
            length++;
        }
        return 0;
    }
    public List<String> neighbors(String word){
        char[] chars = word.toCharArray();
        List<String> result = new ArrayList<>();
        for(int i=0;i<chars.length;i++){
            char temp = chars[i];
            for(char c='a';c<='z';c++){
                chars[i] = c;
                String neighbor = new String(chars);
                result.add(neighbor);
            }
            chars[i] = temp;
        }
        return result;
    }
}

// 126
class Solution {
    public List<List<String>> findLadders(String start, String end, List<String> wordList) {
        HashSet<String> dict = new HashSet<>(wordList);
        List<List<String>> res = new ArrayList<>();
        HashMap<String, ArrayList<String>> nodeNeighbors = new HashMap<>();// Neighbors for every node
        HashMap<String, Integer> distance = new HashMap<String, Integer>();// Distance of every node from the start node
        ArrayList<String> solution = new ArrayList<String>();

        dict.add(start);
        bfs(start, end, dict, nodeNeighbors, distance);
        dfs(start, end, dict, nodeNeighbors, distance, solution, res);
        return res;
    }

    // BFS: Trace every node's distance from the start node (level by level).
    private void bfs(String start, String end, Set<String> dict, HashMap<String, ArrayList<String>> nodeNeighbors, HashMap<String, Integer> distance) {

        for (String str : dict)
        nodeNeighbors.put(str, new ArrayList<>());

        Queue<String> queue = new LinkedList<>();
        queue.offer(start);
        distance.put(start, 0);
        boolean foundEnd = false;
        while (!queue.isEmpty()) {
            int count = queue.size();
            for (int i = 0; i < count; i++) {
                String cur = queue.poll();
                int curDistance = distance.get(cur);
                ArrayList<String> neighbors = getNeighbors(cur, dict);

                for (String neighbor : neighbors) {
                    nodeNeighbors.get(cur).add(neighbor);
                    if (!distance.containsKey(neighbor)) {// Check if visited
                        distance.put(neighbor, curDistance + 1);
                        if (end.equals(neighbor))// Found the shortest path
                            foundEnd = true;
                        else
                            queue.offer(neighbor);
                    }
                }
            }

            if (foundEnd)
            break;
        }
    }

    // Find all next level nodes.
    private ArrayList<String> getNeighbors(String node, Set<String> dict) {
        ArrayList<String> res = new ArrayList<>();
        char chs[] = node.toCharArray();

        for (char ch ='a'; ch <= 'z'; ch++) {
            for (int i = 0; i < chs.length; i++) {
                if (chs[i] == ch) continue;
                char old_ch = chs[i];
                chs[i] = ch;
                if (dict.contains(String.valueOf(chs))) {


                    String newWord = String.valueOf(chs);

                    res.add(String.valueOf(chs));
                }
                chs[i] = old_ch;
            }

        }
        return res;
    }

    // DFS: output all paths with the shortest distance.
    private void dfs(String cur, String end, Set<String> dict, HashMap<String, ArrayList<String>> nodeNeighbors, HashMap<String, Integer> distance, ArrayList<String> solution, List<List<String>> res) {
        solution.add(cur);
        if (end.equals(cur)) {
            res.add(new ArrayList<String>(solution));
        } else {
            for (String next : nodeNeighbors.get(cur)) {
                if (distance.get(next) == distance.get(cur) + 1) {
                    dfs(next, end, dict, nodeNeighbors, distance, solution, res);
                }
            }
        }
        solution.remove(solution.size() - 1);
    }
}
